### ISSUE TITLE
View Metrics error

### ISSUE THREAD

**Issue by toddmcdermid** [DaxStudio/DaxStudio#561]
After:
1. Creating a PBIX with mixed-mode tables (DirectQuery and Import)
2. Publishing the model to a Premium workspace on PowerBI service
3. Refreshing the published model
4. Connecting DAX Studio to that Premium workspace
5. Selecting the published model
6. Clicking "View Metrics"

An error message is received:
2021-03-04 8:49:42 AM		Error viewing metrics: One or more errors occurred.
	TOPNSKIP is not supported in DirectQuery mode

Technical Details:
RootActivityId: b67ce764-d4c4-4f1a-a008-13637bcc14ab
Date (UTC): 3/4/2021 4:49:42 PM

The error (after following suggestions in another post to turn on All Queries tracing - TY) appears to result from:
EVALUATE 
CALCULATETABLE (
ROW( 
    "RelationshipId", 30321,
    "MissingKeys", DISTINCTCOUNT ( 'Customer Invoice Lines'[FK_DimTraveller] ),
    "InvalidRows", COUNTROWS('Customer Invoice Lines')
),
ISBLANK( 'Traveller'[SK_DimTraveller] ),
USERELATIONSHIP( 'Customer Invoice Lines'[FK_DimTraveller], 'Traveller'[SK_DimTraveller] )
)

To work around the error, I went into DAX Studio Options and reduced the "Sample referential integrity violations" in the Advanced tab for VertiPaq Analyzer to zero. This allows DAX Studio to collect metrics.

**Comment by GMedlin** [comment]
I am also getting an error when trying to view metrics on these and even import models now. This is since PBI Desktop's Feb update. The fix mentioned above worked for me as well.

**Comment by dgosbell** [comment]
@GMedlin are you able to share an example import model with some dummy data that shows this issue? I'm not able to reproduce this with pure import mode models. It seems strange to get an error about Direct Query mode if you are using an import mode only model. 

We are adding a retry mechanism that should enable View Metrics to work even when it has problems like this with the RI violation samples.

**Comment by dgosbell** [comment]
@GMedlin yes this sounds like a different issue.  That "An unexpected exception occurred" is coming from the Tabular engine and the SQL dumper windows means the engine actually crashed. Have you seen issue #459?  This sounds similar, but I'm not sure if the people hitting that Microsoft bug are seeing crashes that trigger SQL Dumper. If you go into Help > About in DAX Studio can you open the link to the log folder and see if the most recent log file contains anymore detail about the crash?

**Comment by dgosbell** [comment]
hmm, that does not really add much to what we know already. As far as View Metrics is concerned the original source of the data is irrelevant as we only look at the tabular storage. And local vs remote should not matter too much (although indirectly the amount of available RAM could be an issue in #459 since it appears to be caching related). Do you have SQL Profiler installed? can you try running a trace listening for error events and see if that captures any more information?

**Comment by GMedlin** [comment]
@dgosbell - Sure! Haven't used it before, but it is installed. Here's the error that triggered when I clicked View Metrics in DAX Studio.

> An unexpected error occurred (file 'xmtwdatacache.cpp', line 2719, function 'XMWindowingIterator::Init').
Query Text:
<pii>EVALUATE UNION(CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'Requisition Measures' ),
    "RelationshipId", 6873, 
    "MissingValue", 'Requisition Measures'[Approved Date] 
))),
ISBLANK( 'Approved Date'[Approved Date] ),
USERELATIONSHIP( 'Requisition Measures'[Approved Date], 'Approved Date'[Approved Date] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'Requisition Measures' ),
    "RelationshipId", 14549, 
    "MissingValue", 'Requisition Measures'[glaccountkey] 
))),
ISBLANK( 'GL Accounts'[glaccountkey] ),
USERELATIONSHIP( 'Requisition Measures'[glaccountkey], 'GL Accounts'[glaccountkey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Measures' ),
    "RelationshipId", 14552, 
    "MissingValue", 'PO Measures'[glaccountkey] 
))),
ISBLANK( 'GL Accounts'[glaccountkey] ),
USERELATIONSHIP( 'PO Measures'[glaccountkey], 'GL Accounts'[glaccountkey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Measures' ),
    "RelationshipId", 46917, 
    "MissingValue", 'PO Measures'[adherencekey] 
))),
ISBLANK( 'Adherence'[adherencekey] ),
USERELATIONSHIP( 'PO Measures'[adherencekey], 'Adherence'[adherencekey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61355, 
    "MissingValue", 'PO Balance Measures'[adherencekey] 
))),
ISBLANK( 'Adherence'[adherencekey] ),
USERELATIONSHIP( 'PO Balance Measures'[adherencekey], 'Adherence'[adherencekey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61358, 
    "MissingValue", 'PO Balance Measures'[regionkey] 
))),
ISBLANK( 'Geography'[regionkey] ),
USERELATIONSHIP( 'PO Balance Measures'[regionkey], 'Geography'[regionkey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61361, 
    "MissingValue", 'PO Balance Measures'[systeml4key] 
))),
ISBLANK( 'System L4s'[systeml4key] ),
USERELATIONSHIP( 'PO Balance Measures'[systeml4key], 'System L4s'[systeml4key] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61364, 
    "MissingValue", 'PO Balance Measures'[glaccountkey] 
))),
ISBLANK( 'GL Accounts'[glaccountkey] ),
USERELATIONSHIP( 'PO Balance Measures'[glaccountkey], 'GL Accounts'[glaccountkey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61367, 
    "MissingValue", 'PO Balance Measures'[supplierkey] 
))),
ISBLANK( 'Suppliers'[supplierkey] ),
USERELATIONSHIP( 'PO Balance Measures'[supplierkey], 'Suppliers'[supplierkey] )
),CALCULATETABLE ( 
    TOPN ( 3, DISTINCT ( SELECTCOLUMNS (
        TOPNSKIP ( 30, 0, 'PO Balance Measures' ),
    "RelationshipId", 61370, 
    "MissingValue", 'PO Balance Measures'[sourcesystemkey] 
))),
ISBLANK( 'Source Systems'[sourcesystemkey] ),
USERELATIONSHIP( 'PO Balance Measures'[sourcesystemkey], 'Source Systems'[sourcesystemkey] )
))</pii>

**Comment by dgosbell** [comment]
@GMedlin - so you are definitely hitting the Microsoft bug in issue #459 not the DirectQuery / Composite model issue reported here

**Comment by GMedlin** [comment]
@dgosbell  - OK, thanks for looking into it!
